### PR TITLE
Allow other merge type policies in the repos policy command

### LINKED ISSUE #718: [Feature Request] az repos policy merge-strategy create --use-basic-merge --use-rebase-and-fast-forward --use-rebase-with-merge-commit
**Is your feature request related to a problem? Please describe.**

The web UI of Azure DevOps allows to permit 4 merge strategies in repository branch policies:
* basic merge
* squash merge
* rebase and fast-forward
* rebase with merge commit

The azure-devops-cli-extension only supports configuring squash merges through `az repos policy merge-strategy create --use-squash-merge`.

**Describe the solution you'd like**

`az repos policy merge-strategy create` understands the additional arguments `--use-basic-merge`, `--use-rebase-and-fast-forward` and `--use-rebase-with-merge-commit`

### PR DESCRIPTION
**Pull request by atbagga**
Please make sure the code is following contribution guidelines in CONTRIBUTING.md

 - [ ] : This PR has a corresponding issue open in the Repository. Fixes #718 
 - [ ] : Approach is signed off on the issue.

### CODE REVIEW

File: azure-devops/azext_devops/dev/repos/arguments.py [hunk 1]
```diff
@@ -64,7 +64,21 @@ def load_code_arguments(self, _):
 
     with self.argument_context('repos policy merge-strategy') as context:
         context.argument('use_squash_merge', arg_type=get_three_state_flag(),
-                         help='Whether to squash merge always.')
+                         help='Whether to squash merge always. '
+                         'This option does not work for allowing other merge types.',
+                         deprecate_info=context.deprecate(redirect="--allow-squash"))
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/repos/arguments.py]: How this is shown.. 

image

**Comment by atbagga** [comment]
ToDo: Update recording tests to test newer options. 
Current test will be good enough to test the back compat scenario for the deprecated option.

File: azure-devops/azext_devops/dev/repos/policy.py [hunk 2]
```diff
@@ -215,42 +215,76 @@ def update_policy_required_reviewer(policy_id,
...
+        for i, value in enumerate(param_value_array):
+            if value is None:
+                param_value_array[i] = False
+        # API does not fail but the update is rejected if the last setting is being set to false.
+        # So this check prevents it from client side.
+        if not [i for i, value in enumerate(param_value_array) if value]:
+            raise CLIError("Atleast one merge type must be enabled.")
+    else:
+        param_name_array = ['useSquashMerge']
+        param_value_array = [use_squash_merge]
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/repos/policy.py]: use_squash_merge if use_squash_merge is not None else current_setting.get('useSquashMerge', None) ?

Comment (atbagga) [reply azure-devops/azext_devops/dev/repos/policy.py]: Good catch!

File: azure-devops/azext_devops/dev/repos/policy.py [hunk 3]
```diff
@@ -215,42 +215,108 @@ def update_policy_required_reviewer(policy_id,
...
+        current_setting.get('allowRebaseMerge', None),
+        current_setting.get('allowNoFastForward', None)]
+    is_new_merge_type_update = (allow_squash is not None or
+        allow_no_fast_forward is not None or
+        allow_rebase is not None or
+        allow_rebase_merge is not None)
+
+    if is_new_merge_type_update:
+        if use_squash_merge is not None:
+            raise CLIError("Cannot use new options with the deprecated param --use-squash-merge")
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/repos/policy.py]: ToDo: Review error text.

File: azure-devops/azext_devops/dev/repos/policy.py [hunk 4]
```diff
@@ -215,42 +215,104 @@ def update_policy_required_reviewer(policy_id,
...
     current_policy = policy_client.get_policy_configuration(project=project, configuration_id=policy_id)
-    param_name_array = ['useSquashMerge']
-
     current_setting = current_policy.settings
     current_scope = current_policy.settings['scope'][0]
-
-    param_value_array = [
-        use_squash_merge if use_squash_merge is not None else current_setting.get('useSquashMerge', None)
-    ]
+    current_setting_new_merge_values_array = [
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/repos/policy.py]: @gauravsaralMs @ishitam8 Please review this file again. I didn't handle update in a back compat way earlier. 
Also added tests to check deprecated option, new options and both together.

**Comment by mrooths** [comment]
Hello, When should we expect this functionality to be available?

**Pull request merged**